lib/collection/src/collection_manager/holders/segment_holder.rs

```diff
@@ -1296,73 +1284,58 @@ impl<'s> SegmentHolder {
-                heap.push(DedupPoint {
-                    segment_id,
-                    point_id: next_point_id,
-                });
+            .kmerge_by(|a, b| a.point_id < b.point_id)
+            .chunk_by(|entry| entry.point_id);
+
+        let mut points = Vec::new();
+        let mut points_to_remove: HashMap<SegmentId, Vec<PointIdType>> = HashMap::new();
+
+        for (_, group_iter) in &point_group_iter {
+            // Fill buffer with points of current chunk, need at least 2 points to deduplicate
+            points.clear();
+            points.extend(group_iter);
+            if points.len() < 2 {
+                continue;
             }
 
-            if last_point_id_opt == Some(point_id) {
-                let last_segment_id = last_segment_id_opt.unwrap();
-
-                let point_version = locked_segments[&segment_id].point_version(point_id);
-                let last_point_version = if let Some(last_point_version) = last_point_version_opt {
-                    last_point_version
-                } else {
-                    let version = locked_segments[&last_segment_id].point_version(point_id);
-                    last_point_version_opt = Some(version);
-                    version
-                };
-
-                // choose newer version between point_id and last_point_id
-                if point_version < last_point_version {
-                    log::debug!("Selected point {point_id} in segment {segment_id} for deduplication (version {point_version:?} versus {last_point_version:?} in segment {last_segment_id})");
-
-                    points_to_remove
-                        .entry(segment_id)
-                        .or_default()
-                        .push(point_id);
-                } else {
-                    log::debug!("Selected point {point_id} in segment {last_segment_id} for deduplication (version {last_point_version:?} versus {point_version:?} in segment {segment_id})");
-
-                    points_to_remove
-                        .entry(last_segment_id)
-                        .or_default()
-                        .push(point_id);
+            // Attach point version to
+            for dedup_point in &mut points {
+                dedup_point.version =
+                    locked_segments[&dedup_point.segment_id].point_version(dedup_point.point_id);
+            }
 
-                    last_point_id_opt = Some(point_id);
-                    last_segment_id_opt = Some(segment_id);
-                    last_point_version_opt = Some(point_version);
-                }
-            } else {
-                last_point_id_opt = Some(point_id);
-                last_segment_id_opt = Some(segment_id);
-                last_point_version_opt = None;
+            // Sort points from highest to lowest version
+            points.sort_unstable_by(|a, b| b.version.cmp(&a.version));
+
+            // Keep the first point, remove all others which are older
```

Comment:
Since we don't actually check the other's versions, we assume all other ones are older.

Just to double check, it is not possible to have two points with the same version, is it?

Reply:
> Since we don't actually check the other's versions, we assume all other ones are older.

Yes, but only because they have an older version number. It's the only think that can tell us what point is newer, we cannot rely on any other data to tell us this.

> Just to double check, it is not possible to have two points with the same version, is it?

It is possible, though I'm not sure if it'll ever happen in practice with normal usage.

But even if it does happen, we can assume picking either of the points (having the same version) is fine. :+1:

Reply:
Sorry, I selected the wrong line. Still the questions are answered, but my concern was the operation of removing `[1..]`

Reply:
Check!

Yeah, so it should be fine because:
- index 0 is always the point we want to keep, we consider that to be the highest
- there are always at least 2 points in that loop